File: sidecar/app/tracing.py
"""OpenTelemetry bootstrap for the sidecar.

Spans go to the Jaeger collector over OTLP/HTTP
(http://jaeger:4318/v1/traces by default). The gRPC *aio server* instrumentor
extracts the W3C traceparent the Node gateway injects into call metadata, so
sidecar spans join the gateway's trace instead of starting a new one.

`instrument()` must run before `grpc.aio.server()` is constructed: the
instrumentor works by patching that factory to install its interceptor.
"""

from __future__ import annotations

import logging

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.grpc import GrpcAioInstrumentorServer
from opentelemetry.instrumentation.redis import RedisInstrumentor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

from .config import Config

log = logging.getLogger(__name__)

_provider: TracerProvider | None = None


def setup_tracing(config: Config) -> trace.Tracer:
    """Install the tracer provider and gRPC/Redis instrumentation."""
    global _provider

    resource = Resource.create(
        {
            "service.name": config.service_name,
            "service.version": config.service_version,
            "deployment.environment.name": config.environment,
        }
    )

    exporter = OTLPSpanExporter(endpoint=config.otlp_traces_endpoint, timeout=10)

    # A short delay keeps /trace-test's Jaeger polling from having to wait out
    # the default 5 s batch window on the sidecar side.
    _provider = TracerProvider(resource=resource)
    _provider.add_span_processor(
        BatchSpanProcessor(
            exporter,
            schedule_delay_millis=500,
            max_export_batch_size=512,
            max_queue_size=2048,
        )
    )
    trace.set_tracer_provider(_provider)

    # Patches grpc.aio.server(); must happen before the server is created.
    GrpcAioInstrumentorServer().instrument()
    RedisInstrumentor().instrument()

    log.info("tracing enabled: exporting to %s", config.otlp_traces_endpoint)
    return trace.get_tracer(config.service_name, config.service_version)


def force_flush(timeout_millis: int = 5000) -> None:
    if _provider is not None:
        _provider.force_flush(timeout_millis)


def shutdown_tracing() -> None:
    if _provider is not None:
        _provider.shutdown()

File: api/src/tracing.ts
/**
 * OpenTelemetry bootstrap for the gateway.
 *
 * MUST be imported before fastify/@grpc/grpc-js so the instrumentations can
 * patch those modules — see the import order at the top of src/server.ts.
 *
 * Spans are exported over OTLP/HTTP to the Jaeger collector
 * (http://jaeger:4318/v1/traces by default). The gRPC instrumentation injects
 * W3C traceparent metadata on outgoing calls, which the Python sidecar's
 * server interceptor extracts, so one HTTP request produces a single trace
 * spanning both services.
 */
import { diag, DiagConsoleLogger, DiagLogLevel, trace, type Tracer } from "@opentelemetry/api";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { registerInstrumentations } from "@opentelemetry/instrumentation";
import { FastifyInstrumentation } from "@opentelemetry/instrumentation-fastify";
import { GrpcInstrumentation } from "@opentelemetry/instrumentation-grpc";
import { HttpInstrumentation } from "@opentelemetry/instrumentation-http";
import { Resource } from "@opentelemetry/resources";
import { BatchSpanProcessor } from "@opentelemetry/sdk-trace-base";
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";
import {
  ATTR_SERVICE_NAME,
  ATTR_SERVICE_VERSION,
} from "@opentelemetry/semantic-conventions";

import { config } from "./config.js";

if (process.env.OTEL_DIAG_LOG_LEVEL) {
  const level =
    DiagLogLevel[process.env.OTEL_DIAG_LOG_LEVEL.toUpperCase() as keyof typeof DiagLogLevel];
  diag.setLogger(new DiagConsoleLogger(), level ?? DiagLogLevel.INFO);
}

const exporter = new OTLPTraceExporter({
  url: config.otlpTracesUrl,
  // Keep a failed collector from stalling request handling.
  timeoutMillis: 10_000,
  concurrencyLimit: 10,
});

/**
 * Short batch delay: /trace-test force-flushes anyway, but a small window keeps
 * ordinary traffic visible in the Jaeger UI within a second or so.
 */
const spanProcessor = new BatchSpanProcessor(exporter, {
  scheduledDelayMillis: 500,
  maxExportBatchSize: 512,
  maxQueueSize: 2048,
  exportTimeoutMillis: 10_000,
});

const provider = new NodeTracerProvider({
  resource: new Resource({
    [ATTR_SERVICE_NAME]: config.serviceName,
    [ATTR_SERVICE_VERSION]: config.serviceVersion,
    "deployment.environment.name": config.environment,
  }),
  spanProcessors: [spanProcessor],
});

provider.register();

registerInstrumentations({
  tracerProvider: provider,
  instrumentations: [
    new HttpInstrumentation({
      // The readiness probe and Jaeger query polling would otherwise bury the
      // real traces under health-check noise.
      ignoreIncomingRequestHook: (req) => {
        const url = req.url ?? "";
        return url.startsWith("/healthz") || url.startsWith("/readyz");
      },
      ignoreOutgoingRequestHook: (options) => {
        const path = typeof options.path === "string" ? options.path : "";
        return path.startsWith("/api/traces") || path.startsWith("/api/services");
      },
    }),
    new FastifyInstrumentation(),
    new GrpcInstrumentation(),
  ],
});

/** Tracer for the gateway's own manual spans. */
export const tracer: Tracer = trace.getTracer(config.serviceName, config.serviceVersion);

/**
 * Push everything currently queued to the collector.
 *
 * /trace-test calls this before querying Jaeger: without it the span for the
 * request being verified would still be sitting in the batch processor and the
 * lookup would race the 500 ms export tick.
 */
export async function flushTraces(): Promise<void> {
  await provider.forceFlush();
}

/** Flush and tear down the exporter during graceful shutdown. */
export async function shutdownTracing(): Promise<void> {
  await provider.shutdown();
}
